### ISSUE TITLE
Question for solving overfitting

### ISSUE THREAD

**Issue by jackyhans1** [kjy7567/speech_emotion_recognition_from_log_Mel_spectrogram_using_vertically_long_patch#5]
Hello. First of all, I'd like to express my gratitude for your research. It was truly an interesting paper. I have implemented the code based on your paper, but I'm struggling to solve the overfitting issue. I am reaching out to see if I could get some help with this.

To solve overfitting issue, i've set dropout , emb_dropout = 0.5 , weight_deacy = 1e-3. And I also applied data augmentation but this didn't solve the problem.

Could I get some help with setting the hyperparameters? Additionally, I'm only using Crema-D dataset. 

After reading your paper, I realized you are Korean. I'm Korean too, so please feel free to respond in Korean!

```
teacher_net = Teacher(
    image_size=(128, 128),
    patch_size=(128, 1),
    num_classes=6,
    dim=256,
    depth=6,
    heads=5,
    mlp_dim=256,
    channels=1,
    dropout=0.5,
    emb_dropout=0.5
).to(device)
```

```
student_net = Student(
    image_size=(128, 128),
    patch_size=(128, 1),
    num_classes=6,
    dim=256,
    depth=3,
    heads=5,
    mlp_dim=256,
    channels=1,
    dropout=0.5,
    emb_dropout=0.5
).to(device)
```

```
optimizer_teacher = optim.Adam(teacher_net.parameters(), lr=1e-4, weight_decay=1e-3)
optimizer_student = optim.Adam(student_net.parameters(), lr=1e-4, weight_decay=1e-3)
```

```
# 노이즈 추가 함수
def add_noise(y, noise_factor=0.007):
    noise = np.random.randn(len(y))
    return y + noise_factor * noise
```

```
# 시간 이동 함수
def time_shift(y, shift_max=0.2, sr=16000, min_shift=0.05):
    shift = np.random.randint(sr * min_shift, sr * shift_max)
    direction = np.random.choice([-1, 1])
    shift = shift * direction
    return np.roll(y, shift)
```

```
# 속도 변화 함수
def speed_perturbation(y):
    speed_factor = np.random.choice([0.8, 0.9, 1.1, 1.2])
    return librosa.effects.time_stretch(y.astype(np.float32), rate=speed_factor)
```
    

<img width="736" alt="스크린샷 2024-11-16 오후 6 20 48" src="https://github.com/user-attachments/assets/7d6f2a67-e45b-4a00-a5a4-6a2493b07bba">

**Comment by kjy7567** [comment]
Hello, jackyhans1.
I'm Korean, but I'll write in English so that other people also refer to it :)

First, I would like to know if you considered dividing the trainset and testset into 80%:20% equally for all emotion classes.
For all emotion classes, equal separation is possible through the following code:
```python
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.2, stratify=y_data) 
```

We did not use any dropouts.

We augmented the data using the audiomentation library.
https://github.com/iver56/audiomentations

Sorry, detailed parameters are not available to answer now.

**Comment by jackyhans1** [comment]
Thank you for your fast comment! Yes I divided train set and test set by using stratify. 
It's surprising you didn't use any dropouts. 
Maybe I should try using the audiomentation library for data augmentation.
You've mentioned in the paper that you used noise perturbation, time shifting, and speed perturbation for data augmentation .
I'm wondering if you used every type of data augmentation for each data in train set , test set ending up with 4 times of original dataset.

**Comment by kjy7567** [comment]
Yes, that's correct.
We use all types of augmentation, so the total dataset size is 4 times larger than before.

**Comment by jackyhans1** [comment]
Thank you for your help.
I've got a last question!
Since you said you didn't use any dropout , did you put dropout & emb_dropout as dropout=0.0, emb_dropout=0.0?
Additionally I'm curious if you used weight_decay for weight regularization.
Once again, Thank you for helping me out!

**Comment by kjy7567** [comment]
I set dropout, emb_dropout = 0
Weight decay was not used.
Good luck :)

**Comment by jackyhans1** [comment]
Thank you.
Have a good one
